cmd/preferredimports/preferredimports.go

```diff
@@ -0,0 +1,202 @@
+		a.failed = true
+		return
+	}
+
+	for _, p := range fs {
+		// returns first error, but a.handleError deals with it
+		files := a.filterFiles(p.Files)
+		for _, file := range files {
+			for _, imp := range file.Imports {
+				importPath := strings.Replace(imp.Path.Value, "\"", "", -1)
+				importName := ""
+				if imp.Name != nil {
+					importName = imp.Name.Name
+				}
+				if alias, ok := aliases[importPath]; ok {
+					if alias != importName {
+						_, _ = fmt.Fprintf(os.Stderr, "%sERROR wrong alias for import \"%s\" should be %s in directory %s\n", logPrefix, importPath, alias, dir)
+						a.failed = true
+						return
+					}
+				}
+			}
+		}
+	}
+}
+
+// filterFiles restricts a list of files to only those that should be built by
+// the current platform. This includes both build suffixes (_windows.go) and build
+// tags ("// +build !linux" at the beginning).
+func (a *analyzer) filterFiles(fs map[string]*ast.File) []*ast.File {
+	files := []*ast.File{}
+	for _, f := range fs {
+		fpath := a.fset.File(f.Pos()).Name()
+		dir, name := filepath.Split(fpath)
+		matches, err := a.ctx.MatchFile(dir, name)
+		if err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "%sERROR reading %s: %s\n", logPrefix, fpath, err)
+			a.failed = true
+			continue
+		}
+		if matches {
+			files = append(files, f)
+		}
+	}
+	return files
+}
+
+type collector struct {
+	dirs []string
+}
+
+// handlePath walks the filesystem recursively, collecting directories,
+// ignoring some unneeded directories (hidden/vendored) that are handled
+// specially later.
+func (c *collector) handlePath(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		// Ignore hidden directories (.git, .cache, etc)
+		if len(path) > 1 && path[0] == '.' ||
+			// Staging code is symlinked from vendor/k8s.io, and uses import
+			// paths as if it were inside of vendor/. It fails typechecking
+			// inside of staging/, but works when typechecked as part of vendor/.
+			path == "staging" ||
+			// OS-specific vendor code tends to be imported by OS-specific
+			// packages. We recursively typecheck imported vendored packages for
+			// each OS, but don't typecheck everything for every OS.
+			path == "vendor" ||
+			path == "_output" ||
+			// This is a weird one. /testdata/ is *mostly* ignored by Go,
+			// and this translates to kubernetes/vendor not working.
+			// edit/record.go doesn't compile without gopkg.in/yaml.v2
+			// in $GOSRC/$GOROOT (both typecheck and the shell script).
+			path == "pkg/kubectl/cmd/testdata/edit" {
+			return filepath.SkipDir
+		}
+		c.dirs = append(c.dirs, path)
+	}
+	return nil
+}
+
+func main() {
+	flag.Parse()
+	args := flag.Args()
+
+	if len(args) == 0 {
+		args = append(args, ".")
+	}
+
+	c := collector{}
+	for _, arg := range args {
+		err := filepath.Walk(arg, c.handlePath)
+		if err != nil {
+			log.Fatalf("Error walking: %v", err)
+		}
+	}
+	sort.Strings(c.dirs)
```

Comment:
it will be a lot of errors if scan all the files in k/k when we add a new rule. We could have something like this:

```go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)

	files := []string{}
	for {
		file, err := reader.ReadString('\n')
		if err != nil {
			if err == io.EOF {
				break
			}

			panic(err)
		}

		files = append(files, file)
	}

	fmt.Println(files)
}

```

Reply:
I don't want to optimize this right now as the time taken is pretty small.